lib/r2d2/google_pay_token.rb

```diff
@@ -33,22 +39,98 @@ def decrypt(private_key_pem)
     private
 
     def verify_and_parse_message
+      case protocol_version
+      when 'ECv1'
+        verify_and_parse_message_ecv1
+      when 'ECv2'
+        verify_and_parse_message_ecv2
+      else
+        raise ArgumentError, "unknown protocolVersion #{protocol_version}"
+      end
+    end
+
+    def verify_and_parse_message_ecv1
       digest = OpenSSL::Digest::SHA256.new
       signed_bytes = to_length_value(
         'Google',
         recipient_id,
         protocol_version,
         signed_message
       )
-      verified = verification_keys['keys'].any? do |key|
-        next if key['protocolVersion'] != protocol_version
 
-        ec = OpenSSL::PKey::EC.new(Base64.strict_decode64(key['keyValue']))
-        ec.verify(digest, Base64.strict_decode64(signature), signed_bytes)
-      end
+      verified = valid_key_signatures?(
+        verification_keys,
+        [signature],
+        signed_bytes
+      )
 
       raise SignatureInvalidError unless verified
       JSON.parse(signed_message)
     end
+
+    def verify_and_parse_message_ecv2
+      raise SignatureInvalidError, 'intermediate certificate is expired' if intermediate_key_expired?
+      raise SignatureInvalidError, 'no valid signature of intermediate key' unless root_key_signature_verified?
+      raise SignatureInvalidError, 'signature of signedMessage does not match' unless intermediate_key_signature_verified?
+
+      JSON.parse(signed_message)
+    end
+
+    ### ECv2 Methods ###
+    def root_key_signature_verified?
+      intermediate_signatures = intermediate_signing_key['signatures']
+      signed_key_signature = ['Google', protocol_version, intermediate_signing_key['signedKey']].map do |str|
+        [str.length].pack('V') + str
```

Comment:
In looking over the spec again, I think it details that the `sender`, `protocol_version` and `signed_key` need to be concatenated with "The byte length of each component as 4 bytes in little-endian format.".  I believe the "V" arg to ruby's `Array#pack` method gets us that. Is that your read too?

I'm still experimenting with changing the `Util#to_length_value` method used in ECv1, because it would definitely be nice to have some coherence around way in which we are constructing the byte strings between the two protocol versions.

Reply:
Ah yes, looking again that is what is implemented in Tink. Thanks for calling attention to it again.